### PR TITLE
fix(oncall): sort labels to prevent perpetual drift

### LINKED ISSUE #2559: grafana_oncall_integration: dynamic_labels causes perpetual drift due to TypeList ordering
## Terraform Version

OpenTofu 1.11.5 (also reproducible with Terraform)

## Terraform Grafana Provider Version

Tested with the latest available provider version.

## Affected Resource(s)

- `grafana_oncall_integration`

## Description

The `dynamic_labels` attribute on `grafana_oncall_integration` is defined as `schema.TypeList` in `internal/resources/oncall/resource_integration.go`. Because `TypeList` is ordered, Terraform tracks elements by index position. However, the Grafana OnCall API returns labels in a non-deterministic order (likely by internal ID or creation time), which differs from the order Terraform sends them in.

This causes **perpetual drift** on every `terraform plan` — even immediately after a successful `terraform apply`. The plan shows labels being removed from one position and re-added at another, despite the actual label set being identical.

### Example plan output (immediately after apply)

```
~ dynamic_labels = [
    - {
        - "id"    = "service_name"
        - "key"   = "service_name"
        - "value" = "{{ payload.commonLabels.service_name }}"
      },
      {
          "id"    = "namespace"
          "key"   = "namespace"
          "value" = "{{ payload.commonLabels.namespace }}"
      },
      # (1 unchanged element hidden)
      {
          "id"    = "project"
          "key"   = "project"
          "value" = "{{ payload.commonLabels.project }}"
      },
    + {
        + "id"    = "service_name"
        + "key"   = "service_name"
        + "value" = "{{ payload.commonLabels.service_name }}"
      },
      {
          "id"    = "severity"
          "key"   = "severity"
          "value" = "{{ payload.commonLabels.severity }}"
      },
      # (1 unchanged element hidden)
  ]
```

The label `service_name` is removed from position 0 and re-added at position 3. The actual content is identical — only the order differs.

### Root Cause

In `internal/resources/oncall/resource_integration.go`:

```go
"dynamic_labels": {
    Type: schema.TypeList,
    // ...
},
```

The `flattenLabels` function preserves the API's return order:

```go
func flattenLabels(labels []*onCallAPI.Label) []map[string]string {
    flattenedLabels := make([]map[string]string, 0, 1)
    for _, l := range labels {
        flattenedLabels = append(flattenedLabels, map[string]string{
            "id":    l.Key.Name,
            "key":   l.Key.Name,
            "value": l.Value.Name,
        })
    }
    return flattenedLabels
}
```

Since the API returns labels in a different order than what was sent, and `TypeList` is order-sensitive, every plan detects drift.

## Expected Behavior

`terraform plan` should show no changes after a successful `terraform apply` when the set of dynamic labels has not changed.

## Proposed Fix

Change `dynamic_labels` (and likely `labels` too, since they share the same pattern) from `schema.TypeList` to `schema.TypeSet` with an appropriate hash function. This makes the attribute order-insensitive, which matches the actual semantics — label order is not meaningful in Grafana OnCall.

Alternatively, sort the labels by key in `flattenLabels` before setting the state, so the order is deterministic regardless of API response ordering.

## Workaround

The only current workaround is using `lifecycle { ignore_changes = [dynamic_labels] }`, which prevents Terraform from managing label changes entirely.

## Steps to Reproduce

1. Create a `grafana_oncall_integration` with 2+ dynamic labels
2. Run `terraform apply`
3. Run `terraform plan` immediately after
4. Observe that the plan shows changes (label reordering)
5. Apply again, plan again — drift persists indefinitely

### PR DESCRIPTION
**Pull request by derrix060**
## Summary

- Fix perpetual drift on `grafana_oncall_integration` `labels` and `dynamic_labels` caused by the OnCall API returning labels in non-deterministic order
- Sort labels by key in `flattenLabels` using `sort.SliceStable` to produce a deterministic order regardless of API response ordering
- Add unit tests covering empty, single, sorted, unsorted, and reverse-sorted label inputs

Closes #2559

## Test plan

- [x] Unit tests pass (`go test ./internal/resources/oncall/ -run TestFlattenLabels -v`)
- [x] Acceptance tests with 2+ labels show no drift after apply

### CODE REVIEW

File: internal/resources/oncall/resource_integration.go [hunk 1]
```diff
@@ -851,5 +852,13 @@ func flattenLabels(labels []*onCallAPI.Label) []map[string]string {
 		})
 	}
 
+	// Sort labels by key to produce a deterministic order, preventing
+	// perpetual drift when the API returns labels in a different order.
+	// Use SliceStable because label keys are not guaranteed to be unique
+	// (no uniqueness constraint exists in the provider schema or API client).
```
Comment (derrix060) [inline internal/resources/oncall/resource_integration.go]: Image|

Even though there is a constraint in the UI, but I prefered to be safe here than sorry

File: internal/resources/oncall/resource_integration.go [hunk 2]
```diff
@@ -851,5 +852,20 @@ func flattenLabels(labels []*onCallAPI.Label) []map[string]string {
...
+	// so we add value and id as tie-breakers to ensure a total ordering.
+	sort.SliceStable(flattenedLabels, func(i, j int) bool {
+		if flattenedLabels[i]["key"] != flattenedLabels[j]["key"] {
+			return flattenedLabels[i]["key"] < flattenedLabels[j]["key"]
+		}
+		if flattenedLabels[i]["value"] != flattenedLabels[j]["value"] {
+			return flattenedLabels[i]["value"] < flattenedLabels[j]["value"]
+		}
+		return flattenedLabels[i]["id"] < flattenedLabels[j]["id"]
+	})
```
[the remark this answers is not in the document]
Comment (derrix060) [reply internal/resources/oncall/resource_integration.go]: This is a breaking change.  I'm happy with it, but I'm defering to the repo mantainers

**Comment by gerardocorea** [comment]
@derrix060 thank you for your contribution. A recently closed PR https://github.com/grafana/terraform-provider-grafana/pull/2536 should have addressed this issue. Due to this I will be closing out this PR. thank you.
